Remove optional for veiw_fn during View Tracking

ejguan opened the pull request:
Fixes #49257

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50067 Remove optional for veiw_fn during View Tracking**

Using the `Callgrind` to test the performance.
```python
import torch
import timeit
from torch.utils.benchmark import Timer

timer = Timer("x.view({100, 5, 20});", setup="torch::Tensor x = torch::ones({10, 10, 100});", language="c++", timer=timeit.default_timer)
res = timer.collect_callgrind(number=10)
```
### Nightly
```python
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fdd49d61c70>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        41850                      41850
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Current
```python
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7fdcf322d1c0>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42020                      42020
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Compare
There are 170 instructions added
```python
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7fdd427df7f0>
    970  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
    240  ???:torch::autograd::ViewInfo::~ViewInfo()
    180  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, std::function<at::Tensor (at::Tensor const&)>)
    130  ???:torch::autograd::make_variable_differentiable_view(at::Tensor const&, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta, bool)
    105  /tmp/benchmark_utils_jit_build_66d13e5035484028ab02edcb72545e37/timer_cpp_2208139565547260839/timer_src.cpp:main
    100  ???:std::function<at::Tensor (at::Tensor const&)>::function(std::function<at::Tensor (at::Tensor const&)> const&)
     70  ???:torch::autograd::DifferentiableViewMeta::~DifferentiableViewMeta()
     70  ???:torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta(c10::TensorImpl*, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta)
   -100  ???:c10::optional_base<torch::autograd::ViewInfo>::optional_base(c10::optional_base<torch::autograd::ViewInfo>&&)
   -105  /tmp/benchmark_utils_jit_build_a62a84fbc4cd4f9e91376c1bb3dcb697/timer_cpp_7842929001219208348/timer_src.cpp:main
   -120  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, c10::optional<std::function<at::Tensor (at::Tensor const&)> >)
   -210  ???:c10::optional_base<std::function<at::Tensor (at::Tensor const&)> >::~optional_base()
   -240  ???:c10::optional_base<torch::autograd::ViewInfo>::~optional_base()
   -920  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, c10::optional<std::function<at::Tensor (at::Tensor const&)> >, torch::autograd::CreationMeta, bool)

Total: 170
```

Differential Revision: [D25900495](https://our.internmc.facebook.com/intern/diff/D25900495)

ejguan committed:
Remove optional for veiw_fn during View Tracking

[ghstack-poisoned]

ejguan committed:
Update on "Remove optional for veiw_fn during View Tracking"

Fixes #49257


Using the following script to test the performance.
```python
import torch

x = torch.randn(100, 100, requires_grad=True)
with torch.autograd.profiler.profile() as prof:
    for _ in range(100):
        y = torch.sum(torch.diagonal(x))
        y.backward()

print(prof.key_averages().table(sort_by="self_cpu_time_total"))
```

### Nightly
![image](https://user-images.githubusercontent.com/68879799/103581107-53f5f680-4ea9-11eb-800a-d6570d1c8b6b.png)

### Current
![image](https://user-images.githubusercontent.com/68879799/103581114-5bb59b00-4ea9-11eb-9aa8-038741bc8407.png)

I don't find any significant regression over backward function (using the `view_fn`)

Differential Revision: [D25768847](https://our.internmc.facebook.com/intern/diff/D25768847)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/functions/tensor.cpp:
```diff
@@ -79,8 +79,8 @@ auto CopySlices::apply(variable_list&& inputs) -> variable_list {
-  if (view_fn.has_value()) {
-    auto fn = view_fn.value();
+  if (view_fn) {
+    auto fn = view_fn;
```
Is this still needed now that it is not an optional anymore?

ejguan replied:
Yes, it's needed as `view_fn` can be `nullptr` for CopySlices.
As the [comment](https://github.com/pytorch/pytorch/blob/eee849be8cc6cce5ce867afb7d95aec521b16dce/torch/csrc/autograd/functions/tensor.h#L60-L67) shown, on CPU/GPU, we can use TensorGeometry of both base and view tensors rather than `view_fn` to create CopySlices.
https://github.com/pytorch/pytorch/pull/50067#discussion_r552211994

albanD replied:
Right but why not directly use `view_fn` in the call below? This shallow copy of the function isn't doing anything here?

ejguan replied:
I see. You are right. I should directly call `view_fn`.

albanD reviewed (commented):
This change looks good.

Can you check the perf implication with the new Timer class (https://pytorch.org/docs/master/benchmark_utils.html?highlight=timer#torch.utils.benchmark.Timer) with the instruction count mode that will be much more precise!
You can do somthing like this to save a trace (here for a basic view op, but maybe something else if it is more relevant to your PR):
```python
import torch
from torch.utils.benchmark import Timer

timer = Timer("x.view(-1);", setup="torch::Tensor x = torch::ones({1,2,3});", language="c++")
res = timer.collect_callgrind(number=10)

torch.save(res, "save.pth")
```

And once you do that for master and your branch, you can compare with:
```python
import torch
torch.set_printoptions(linewidth=260)

master_instr = torch.load("master.pth")
fast_view_instr = torch.load("fast_view.pth")

master_instr = master_instr.as_standardized()
fast_view_instr = fast_view_instr.as_standardized()


print(master_instr)
print(fast_view_instr)

delta = fast_view_instr.delta(master_instr).denoise()

print(delta)
```

ejguan commented on torch/csrc/autograd/variable.h:
```diff
@@ -314,9 +314,9 @@ struct TORCH_API ViewInfo {
   /// The "view_func", if provided, should be a function that allows to re-do the view
   /// between "base" and "tensor".
   ViewInfo chain(const Variable & base, const Variable & tensor,
-    c10::optional<std::function<Variable(const Variable&)>> view_func=c10::nullopt) const;
+    std::function<Variable(const Variable&)> view_func=nullptr) const;
```
It can return a `ViewInfo` with `nullptr`

ejguan committed:
Update on "Remove optional for veiw_fn during View Tracking"

Fixes #49257


Using the following script to test the performance.
```python
import torch

x = torch.randn(100, 100, requires_grad=True)
with torch.autograd.profiler.profile() as prof:
    for _ in range(100):
        y = torch.sum(torch.diagonal(x))
        y.backward()

print(prof.key_averages().table(sort_by="self_cpu_time_total"))
```

### Nightly
![image](https://user-images.githubusercontent.com/68879799/103666773-0d54da80-4f43-11eb-9f04-d7b94276f82f.png)

### Current
![image](https://user-images.githubusercontent.com/68879799/103665874-eba72380-4f41-11eb-9843-61fcd801464f.png)

I don't find any significant regression over backward function (using the `view_fn`)

Differential Revision: [D25768847](https://our.internmc.facebook.com/intern/diff/D25768847)

[ghstack-poisoned]

ejguan committed:
Update on "Remove optional for veiw_fn during View Tracking"

Fixes #49257


Using the following script to test the performance.
```python
import torch

x = torch.randn(100, 100, requires_grad=True)
with torch.autograd.profiler.profile() as prof:
    for _ in range(100):
        y = torch.sum(torch.diagonal(x))
        y.backward()

print(prof.key_averages().table(sort_by="self_cpu_time_total"))
```

### Nightly
![image](https://user-images.githubusercontent.com/68879799/103666773-0d54da80-4f43-11eb-9f04-d7b94276f82f.png)

### Current
![image](https://user-images.githubusercontent.com/68879799/103665874-eba72380-4f41-11eb-9843-61fcd801464f.png)

I don't find any significant regression over backward function (using the `view_fn`)

Differential Revision: [D25768847](https://our.internmc.facebook.com/intern/diff/D25768847)

[ghstack-poisoned]

ejguan committed:
Update on "Remove optional for veiw_fn during View Tracking"

Fixes #49257


Using the `Callgrind` to test the performance.
```python
import torch
import timeit
from torch.utils.benchmark import Timer

timer = Timer("x.view({100, 5, 20});", setup="torch::Tensor x = torch::ones({10, 10, 100});", language="c++", timer=timeit.default_timer)
res = timer.collect_callgrind(number=10)
```
### Nightly
```python
torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f7949138c40>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42310                      42310
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Current
```python
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f78f271a580>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42480                      42480
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Compare
There are 170 instructions reduced
```python
torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f7941b7a7c0>
    970  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
    240  ???:torch::autograd::ViewInfo::~ViewInfo()
    180  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, std::function<at::Tensor (at::Tensor const&)>)
    130  ???:torch::autograd::make_variable_differentiable_view(at::Tensor const&, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta, bool)
    105  /tmp/benchmark_utils_jit_build_69e2f1710544485588feeca0719a3a57/timer_cpp_4435526292782672407/timer_src.cpp:main
    100  ???:std::function<at::Tensor (at::Tensor const&)>::function(std::function<at::Tensor (at::Tensor const&)> const&)
     70  ???:torch::autograd::DifferentiableViewMeta::~DifferentiableViewMeta()
     70  ???:torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta(c10::TensorImpl*, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta)
   -100  ???:c10::optional_base<torch::autograd::ViewInfo>::optional_base(c10::optional_base<torch::autograd::ViewInfo>&&)
   -105  /tmp/benchmark_utils_jit_build_2e75f38b553e42eba00523a86ad9aa05/timer_cpp_3360771523810516633/timer_src.cpp:main
   -120  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, c10::optional<std::function<at::Tensor (at::Tensor const&)> >)
   -210  ???:c10::optional_base<std::function<at::Tensor (at::Tensor const&)> >::~optional_base()
   -240  ???:c10::optional_base<torch::autograd::ViewInfo>::~optional_base()
   -920  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, c10::optional<std::function<at::Tensor (at::Tensor const&)> >, torch::autograd::CreationMeta, bool)
```
Differential Revision: [D25768847](https://our.internmc.facebook.com/intern/diff/D25768847)

[ghstack-poisoned]

ejguan committed:
Update on "Remove optional for veiw_fn during View Tracking"

Fixes #49257


Using the `Callgrind` to test the performance.
```python
import torch
import timeit
from torch.utils.benchmark import Timer

timer = Timer("x.view({100, 5, 20});", setup="torch::Tensor x = torch::ones({10, 10, 100});", language="c++", timer=timeit.default_timer)
res = timer.collect_callgrind(number=10)
```
### Nightly
```python
torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f7949138c40>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42310                      42310
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Current
```python
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f78f271a580>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42480                      42480
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Compare
There are 170 instructions reduced
```python
torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f7941b7a7c0>
    970  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
    240  ???:torch::autograd::ViewInfo::~ViewInfo()
    180  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, std::function<at::Tensor (at::Tensor const&)>)
    130  ???:torch::autograd::make_variable_differentiable_view(at::Tensor const&, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta, bool)
    105  /tmp/benchmark_utils_jit_build_69e2f1710544485588feeca0719a3a57/timer_cpp_4435526292782672407/timer_src.cpp:main
    100  ???:std::function<at::Tensor (at::Tensor const&)>::function(std::function<at::Tensor (at::Tensor const&)> const&)
     70  ???:torch::autograd::DifferentiableViewMeta::~DifferentiableViewMeta()
     70  ???:torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta(c10::TensorImpl*, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta)
   -100  ???:c10::optional_base<torch::autograd::ViewInfo>::optional_base(c10::optional_base<torch::autograd::ViewInfo>&&)
   -105  /tmp/benchmark_utils_jit_build_2e75f38b553e42eba00523a86ad9aa05/timer_cpp_3360771523810516633/timer_src.cpp:main
   -120  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, c10::optional<std::function<at::Tensor (at::Tensor const&)> >)
   -210  ???:c10::optional_base<std::function<at::Tensor (at::Tensor const&)> >::~optional_base()
   -240  ???:c10::optional_base<torch::autograd::ViewInfo>::~optional_base()
   -920  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, c10::optional<std::function<at::Tensor (at::Tensor const&)> >, torch::autograd::CreationMeta, bool)
```

[ghstack-poisoned]

ejguan committed:
Update on "Remove optional for veiw_fn during View Tracking"

Fixes #49257


Using the `Callgrind` to test the performance.
```python
import torch
import timeit
from torch.utils.benchmark import Timer

timer = Timer("x.view({100, 5, 20});", setup="torch::Tensor x = torch::ones({10, 10, 100});", language="c++", timer=timeit.default_timer)
res = timer.collect_callgrind(number=10)
```
### Nightly
```python
torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f7949138c40>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42310                      42310
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Current
```python
<torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.CallgrindStats object at 0x7f78f271a580>
x.view({100, 5, 20});
setup: torch::Tensor x = torch::ones({10, 10, 100});
                           All          Noisy symbols removed
    Instructions:        42480                      42480
    Baseline:                0                          0
10 runs per measurement, 1 thread
Warning: PyTorch was not built with debug symbols.
         Source information may be limited. Rebuild with
         REL_WITH_DEB_INFO=1 for more detailed results.
```
### Compare
There are 170 instructions reduced
```python
torch.utils.benchmark.utils.valgrind_wrapper.timer_interface.FunctionCounts object at 0x7f7941b7a7c0>
    970  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, std::function<at::Tensor (at::Tensor const&)>, torch::autograd::CreationMeta, bool)
    240  ???:torch::autograd::ViewInfo::~ViewInfo()
    180  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, std::function<at::Tensor (at::Tensor const&)>)
    130  ???:torch::autograd::make_variable_differentiable_view(at::Tensor const&, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta, bool)
    105  /tmp/benchmark_utils_jit_build_69e2f1710544485588feeca0719a3a57/timer_cpp_4435526292782672407/timer_src.cpp:main
    100  ???:std::function<at::Tensor (at::Tensor const&)>::function(std::function<at::Tensor (at::Tensor const&)> const&)
     70  ???:torch::autograd::DifferentiableViewMeta::~DifferentiableViewMeta()
     70  ???:torch::autograd::DifferentiableViewMeta::DifferentiableViewMeta(c10::TensorImpl*, c10::optional<torch::autograd::ViewInfo>, c10::optional<torch::autograd::ViewInfo>, torch::autograd::CreationMeta)
   -100  ???:c10::optional_base<torch::autograd::ViewInfo>::optional_base(c10::optional_base<torch::autograd::ViewInfo>&&)
   -105  /tmp/benchmark_utils_jit_build_2e75f38b553e42eba00523a86ad9aa05/timer_cpp_3360771523810516633/timer_src.cpp:main
   -120  ???:torch::autograd::ViewInfo::ViewInfo(at::Tensor, c10::optional<std::function<at::Tensor (at::Tensor const&)> >)
   -210  ???:c10::optional_base<std::function<at::Tensor (at::Tensor const&)> >::~optional_base()
   -240  ???:c10::optional_base<torch::autograd::ViewInfo>::~optional_base()
   -920  ???:torch::autograd::as_view(at::Tensor const&, at::Tensor const&, bool, bool, c10::optional<std::function<at::Tensor (at::Tensor const&)> >, torch::autograd::CreationMeta, bool)
```


[ghstack-poisoned]

ejguan commented:
@albanD 
It doesn't seem there is huge benefit in terms of instruction counts by removing optional for function, as shown in the comment.

ejguan commented:
@albanD 
It's quite interesting that It doesn't seem there is benefit in terms of instruction counts by removing optional for function, as shown in the above comment.

albanD commented:
Hi,

Thanks for the benchmark, after discussing with other people, it is expected that the runtime is the same (the compiler is too smart for us): https://godbolt.org/z/W16xos (thanks @swolchok for the godbolt link).
One benefit that is not measured here though is the smaller memory footprint which is always nice!

albanD reviewed (approved):
The updated code looks good to me, thanks!
